doc/developer/design/20210707_qgm_sql_high_level_representation.md

```diff
@@ -0,0 +1,726 @@
+  Note: temporarily during the generation of the query model we could allow subquery quantifiers in `OuterJoin` boxes for
+  subqueries in the `ON`-clause of the outer join, but should push down the subquery to the non-preserving side.
+  Note 2: In QGM there is no distinction between a `LEFT JOIN` and a `RIGHT JOIN`, since that's a concept that belongs
+  only in the AST.
+
+Some of the constraints above are just conventions for making query transformation easier due to having to cover
+fewer cases. The rest are just constructions that don't make sense semantically speaking.
+
+We could make some of these constraints explicit by making the type system enforce them, by, for example
+moving the `quantifiers` set to each of the `BoxType` that can/must have input quantifiers, which are all of
+them but `BaseTable`, `TableFunction` and `Values`.
+
+All boxes have an ordered projection, represented as a vector of columns, defined as:
+
+```rust
+struct Column {
+    expr: Expr,
+    alias: Option<String>,
+}
+```
+
+### Notes on expression representation
+
+Columns have two representations in QGM: base columns and column references. Base columns are only allowed in expressions
+contained in data source operators, specifically in the projection of boxes of type `BaseTable` and `TableFunction`.
+
+```rust
+enum Expr {
+    // ...
+    ColumnReference(ColumnReference),
+    BaseColumn(usize),
+}
+
+struct ColumnReference {
+    quantifier_id: QuantifierId,
+    position: usize,
+}
+```
+
+`ColumnReference` is used everywhere else. A `ColumnReference` may either reference a quantifier of the same
+box that owns the containing expression or a quantifier from some parent box.
+
+The underlying expression behind a column reference can be obtained via a `dereference` method, whose implementation
+could be as follows:
+
+```rust
+impl ColumnReference {
+    fn dereference<'a>(&self, model: &'a Model) -> &'a Expr {
+        let input_box = model
+            .quantifiers
+            .get(&self.quantifier_id)
+            .unwrap()
+            .input_box;
+        &model.boxes.get(&input_box).unwrap().columns[self.position].expr
+    }
+}
+```
+
+Since this proposal uses identifiers instead of pointers, most methods in the implementation `Expr` will need to
+receive a reference to the model as a parameter. For example, a method for determining whether an expression is
+nullable or not may need to dereference a column reference, for which it needs access to the model:
+
+```rust
+impl Expr {
+    fn nullable(&self, model: &Model) -> bool {
+        match self {
+            ...
+            Expr::ColumnReference(c) => c.dereference(model).nullable(model),
+        }
+    }
+}
+```
+
+### Name resolution
+
+As shown above, the query graph already contains almost all the information needed for name resolution. Since the
+query graph is built in a bottom-up manner, we can use the input quantifiers for resolving names within the
```

Comment:
I'm a bit confused about "query graph is built in a bottom-up manner" since the diagrams in this section generally have Box0 at the top. Is it that boxes are initialized from the top down, but the bottom-most box will be fully filled in first?

Does name resolution happen in the same order that quantifiers are added to the model?

Reply:
Once again: exactly. The order the of the creation of the boxes doesn't really matter, but the order in which they become fully filled and added to the model with a quantifier ranging over them is the relevant one. That happens bottom-up.

Note: the boxes that define a name resolution scope are created in a top-down manner. Some other boxes, for example `Union` boxes or boxes for applying column aliases, may be created after its input boxes are created.

I'll try to find a better wording for this.